### ISSUE TITLE
`at(1) >> 4)` caused `call to 'to_string' is ambiguous`

### ISSUE THREAD

**Issue by duffy-corelight** [zeek/spicy#216]
In my build where at() was still yielding uint64, the difference:
```
3302 /usr/local/etc/tmp$ diff Error_2741_non-ambiguous.spicy Error_2741_ambiguous.spicy
26c26
<         print 10 * (*rawTimeAndTimezone.at(1) & 0x0F), (*rawTimeAndTimezone.at(1));
---
>         print 10 * (*rawTimeAndTimezone.at(1) & 0x0F), (*rawTimeAndTimezone.at(1) >> 4);
```
was a change between the code working, and the code producing error:
```
3300 /usr/local/etc/tmp$ spicy/build/bin/spicyc -j Error_2741_non-ambiguous.spicy
2000, 13, 0
0, 16
1-00-00T00:00:00
3301 /usr/local/etc/tmp$ spicy/build/bin/spicyc -j Error_2741_ambiguous.spicy
In file included from tzDatetime.cc:6:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/libhilti.h:5:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/configuration.h:8:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/util.h:30:
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:31:12: error: call to 'to_string' is ambiguous
    return to_string(x, detail::adl::tag{});
           ^~~~~~~~~
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:69:23: note: in instantiation of function template specialization 'hilti::rt::to_string<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    return hilti::rt::to_string(x);
                      ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:81:20: note: in instantiation of function template specialization 'hilti::rt::detail::to_string_for_print<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    return detail::to_string_for_print(x);
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/util.h:379:39: note: in instantiation of function template specialization 'hilti::rt::to_string_for_print<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    std::vector<std::string> x = {rt::to_string_for_print(std::get<Is>(std::forward<T>(tup)))...};
                                      ^
tzDatetime.cc:50:16: note: in instantiation of function template specialization 'hilti::rt::printValues<std::__1::tuple<SafeInt<unsigned long long, hilti::rt::integer::detail::SafeIntException>, SafeInt<char, hilti::rt::integer::detail::SafeIntException> >, nullptr>' requested here
    hilti::rt::printValues(std::make_tuple(hilti::rt::integer::safe<std::uint64_t>{10u} * (*rawTimeAndTimezone.at(hilti::rt::integer::safe<std::uint64_t>{1u}) & hilti::rt::integer::safe<std::uint64_t>{15u}), *rawTimeAndTimezone.at(hilti::rt::integer::safe<std::uint64_t>{1u}) >> hilti::rt::integer::safe<std::uint64_t>{4u}), true);
               ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/bool.h:10:20: note: candidate function
inline std::string to_string(bool x, adl::tag /*unused*/) { return (x ? "True" : "False"); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:19:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:23:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:27:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:31:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:35:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:39:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:43:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint8_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:47:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int8_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:51:20: note: candidate function
inline std::string to_string(uint64_t x, adl::tag /*unused*/) { return fmt("%" PRIu64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:53:20: note: candidate function
inline std::string to_string(int64_t x, adl::tag /*unused*/) { return fmt("%" PRId64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:55:20: note: candidate function
inline std::string to_string(uint32_t x, adl::tag /*unused*/) { return fmt("%" PRIu32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:57:20: note: candidate function
inline std::string to_string(int32_t x, adl::tag /*unused*/) { return fmt("%" PRId32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:59:20: note: candidate function
inline std::string to_string(uint16_t x, adl::tag /*unused*/) { return fmt("%" PRIu16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:61:20: note: candidate function
inline std::string to_string(int16_t x, adl::tag /*unused*/) { return fmt("%" PRId16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:63:20: note: candidate function
inline std::string to_string(uint8_t x, adl::tag /*unused*/) { return fmt("%" PRIu8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:65:20: note: candidate function
inline std::string to_string(int8_t x, adl::tag /*unused*/) { return fmt("%" PRId8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/real.h:25:20: note: candidate function
inline std::string to_string(double x, adl::tag /*unused*/) {
                   ^
1 error generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit tzDatetime to bitcode
[error] spicyc: JIT compilation failed
```

**Comment by rsmmr** [comment]
I believe that the solution in #197 needs another tweak: instead of returning uint8_t when deferencing a bytes iterator, we should return `hilti::rt::integer::safe<uint8_t>`. That should then select the right overload.

**Comment by bbannier** [comment]
@duffy-corelight, can you please include a minimal example which reproduces the issue? I would also be interested to know what version you are running (e.g., from `spicyc --version`).

**Comment by duffy-corelight** [comment]
```
spicy/build/bin/spicyc --version
spicyc v (master ae5968ce)
```

**Comment by duffy-corelight** [comment]
```
module tzDatetime;

 calledtzDatetime(b"13");

function calledtzDatetime(rawTimeAndTimezone: bytes) {
    print 10 * (*rawTimeAndTimezone.at(1) & 0x0F), (*rawTimeAndTimezone.at(1));
    print 10 * (*rawTimeAndTimezone.at(1) & 0x0F), (*rawTimeAndTimezone.at(1) >> 4);
}
```
outputs:
```
spicy/build/bin/spicyc -j Issue_216_shorter.spicy
[error] Issue_216_shorter.spicy:3: call does not match any function: calledtzDatetime()
[error] Issue_216_shorter.spicy:3:   candidate functions:
[error] Issue_216_shorter.spicy:3:   - calledtzDatetime(<const bytes>)
[error] spicyc: aborting after errors
4206 /usr/local/etc/tmp$spicy/build/bin/spicyc -j Issue_216_shorter.spicy
In file included from tzDatetime.cc:6:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/libhilti.h:5:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/configuration.h:8:
In file included from /usr/local/etc/tmp/spicy/hilti/include/hilti/rt/util.h:30:
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:31:12: error: call to 'to_string' is ambiguous
    return to_string(x, detail::adl::tag{});
           ^~~~~~~~~
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:69:23: note: in instantiation of function template specialization 'hilti::rt::to_string<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    return hilti::rt::to_string(x);
                      ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/extension-points.h:81:20: note: in instantiation of function template specialization 'hilti::rt::detail::to_string_for_print<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    return detail::to_string_for_print(x);
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/util.h:379:39: note: in instantiation of function template specialization 'hilti::rt::to_string_for_print<SafeInt<char, hilti::rt::integer::detail::SafeIntException> >' requested here
    std::vector<std::string> x = {rt::to_string_for_print(std::get<Is>(std::forward<T>(tup)))...};
                                      ^
tzDatetime.cc:23:16: note: in instantiation of function template specialization 'hilti::rt::printValues<std::__1::tuple<SafeInt<unsigned long long, hilti::rt::integer::detail::SafeIntException>, SafeInt<char, hilti::rt::integer::detail::SafeIntException> >, nullptr>' requested here
    hilti::rt::printValues(std::make_tuple(hilti::rt::integer::safe<std::uint64_t>{10u} * (*rawTimeAndTimezone.at(hilti::rt::integer::safe<std::uint64_t>{1u}) & hilti::rt::integer::safe<std::uint64_t>{15u}), *rawTimeAndTimezone.at(hilti::rt::integer::safe<std::uint64_t>{1u}) >> hilti::rt::integer::safe<std::uint64_t>{4u}), true);
               ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/bool.h:10:20: note: candidate function
inline std::string to_string(bool x, adl::tag /*unused*/) { return (x ? "True" : "False"); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:19:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:23:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int64_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:27:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:31:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int32_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:35:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:39:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int16_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:43:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<uint8_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:47:20: note: candidate function
inline std::string to_string(hilti::rt::integer::safe<int8_t> x, adl::tag /*unused*/) {
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:51:20: note: candidate function
inline std::string to_string(uint64_t x, adl::tag /*unused*/) { return fmt("%" PRIu64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:53:20: note: candidate function
inline std::string to_string(int64_t x, adl::tag /*unused*/) { return fmt("%" PRId64, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:55:20: note: candidate function
inline std::string to_string(uint32_t x, adl::tag /*unused*/) { return fmt("%" PRIu32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:57:20: note: candidate function
inline std::string to_string(int32_t x, adl::tag /*unused*/) { return fmt("%" PRId32, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:59:20: note: candidate function
inline std::string to_string(uint16_t x, adl::tag /*unused*/) { return fmt("%" PRIu16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:61:20: note: candidate function
inline std::string to_string(int16_t x, adl::tag /*unused*/) { return fmt("%" PRId16, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:63:20: note: candidate function
inline std::string to_string(uint8_t x, adl::tag /*unused*/) { return fmt("%" PRIu8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/integer.h:65:20: note: candidate function
inline std::string to_string(int8_t x, adl::tag /*unused*/) { return fmt("%" PRId8, x); }
                   ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/rt/types/real.h:25:20: note: candidate function
inline std::string to_string(double x, adl::tag /*unused*/) {
                   ^
1 error generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit tzDatetime to bitcode
[error] spicyc: JIT compilation failed
```
then by putting a # character for commenting-out the latter line, 
```
#    print 10 * (*rawTimeAndTimezone.at(1) & 0x0F), (*rawTimeAndTimezone.at(1) >> 4);  
```
it outputs:
```
4207 /usr/local/etc/tmp$spicy/build/bin/spicyc -j Issue_216_shorter.spicy
30, 51
```

**Comment by bbannier** [comment]
This was fixed by 73326179a0b9349b49cedd62a06e5634439a3465 which was part of the fix for #191.
